File: 0856-score-of-parentheses/0856-score-of-parentheses.py
class Solution:
    def scoreOfParentheses(self, s: str) -> int:
        stack=[]
        output=value=0
        for i in s:
            if i=='(':
                stack.append(0)
            else:
                poppedValue=stack.pop()
                value=1 if poppedValue == 0 else poppedValue*2
                if not stack:
                    output+=value
                else:
                    stack[-1]+=value
        return output
    
    
#         k=1
#         score=0
#         openParanthesis=True
#         for p in s:
#             if p=='(':
#                 k*=2
#                 openParanthesis=True
#             else:
#                 k//=2
#                 if openParanthesis:
#                     score+=k
#                     openParanthesis=False
#         return score

        

File: 0856-score-of-parentheses/0856-score-of-parentheses.java
class Solution {
    public int scoreOfParentheses(String s) {
        Stack<Character> stack = new Stack();

        for(char ch : s.toCharArray()) {
            if(ch == '(') {
                stack.push(ch);
            } else {
                // if top of stack is '('
                if(!stack.isEmpty() && stack.peek() == '(') {
                    stack.pop();
                    stack.push('1');
                } else {
                    // if top of stack is a '(char) number'
                    int sum = 0;
                    while(!stack.isEmpty() && stack.peek()!='(') {
                        sum += stack.pop()-'0';
                    }
                    // pop '('
                    stack.pop();
                    stack.push((char)(2*sum+'0'));
                }
            }
        }
        
        int sum = 0;
        while(!stack.isEmpty()) {
            sum += (stack.pop()-'0');
        }
        return sum;
    }
}